libtenzir/builtins/aggregation-functions/min.cpp

```diff
@@ -42,7 +47,94 @@ class min_function final : public aggregation_function {
   std::optional<type_to_data_t<Type>> min_ = {};
 };
 
-class plugin : public virtual aggregation_function_plugin {
+class min_instance final : public aggregation_instance {
+public:
+  using min_t = variant<caf::none_t, int64_t, uint64_t, double, duration>;
+  explicit min_instance(ast::expression expr) : expr_{std::move(expr)} {
+  }
+
+  auto update(const table_slice& input, session ctx) -> void override {
+    if (min_ and std::holds_alternative<caf::none_t>(min_.value())) {
+      return;
+    }
+    auto arg = eval(expr_, input, ctx);
+    if (not type_) {
+      type_ = arg.type;
+    }
+    const auto warn = [&](const auto&) -> min_t {
+      diagnostic::warning("expected `{}`, got `{}`", type_, arg.type)
+        .primary(expr_)
+        .emit(ctx);
+      return caf::none_t{};
+    };
```

Comment:
We unconditionally remember the first type that we evaluated here. However, how do we know that the first type is a correct type? Does this work if I use a string as the first type?

Reply:
We do not know the first type is correct. However, in that case, the diagnostic emitted is from [here](https://github.com/tenzir/tenzir/pull/4542/files#diff-0e78de9e52a72771e3e119cc512883056fa5bf5e0deac6c7bfa02e56770833edR110-R117). So, we effectively consider that the type is correct when used.